lib/logflare/backends/adaptor/clickhouse_adaptor/provisioner.ex

```diff
@@ -0,0 +1,69 @@
+defmodule Logflare.Backends.Adaptor.ClickhouseAdaptor.Provisioner do
+  @moduledoc """
+  Short-lived process that kicks off when spinning up a `ClickhouseAdaptor`.
+  Used to provision resources (tables/views) within ClickHouse.
+
+  It is assumed that the `ClickhouseAdaptor` is running _before_ starting this process.
+  """
+
+  use GenServer
+  use TypedStruct
+  require Logger
+
+  alias Logflare.Backends.Adaptor.ClickhouseAdaptor
+  alias Logflare.Backends.Backend
+  alias Logflare.Source
+
+  typedstruct do
+    field(:source, Source.t())
+    field(:backend, Backend.t())
+  end
+
+  @doc false
+  def child_spec(arg) do
+    %{
+      id: __MODULE__,
+      restart: :transient,
+      start: {__MODULE__, :start_link, [arg]}
+    }
+  end
+
+  @doc false
+  @spec start_link(ClickhouseAdaptor.source_backend_tuple()) :: GenServer.on_start()
+  def start_link({%Source{}, %Backend{}} = args) do
+    GenServer.start_link(__MODULE__, args)
+  end
+
+  @doc false
+  @impl GenServer
+  def init({%Source{} = source, %Backend{} = backend} = args) do
+    state = %__MODULE__{
+      source: source,
```

Comment:
seems like we don't need to store the entire structs in state, can just store the source_id source_token backend_id.
this would consume some memory footprint, though minimal.

Reply:
Minor change made to maintain only those 3 attributes of the structs